modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ImageConvertWorkflowOperationHandler.java

```diff
@@ -0,0 +1,248 @@
+      throw new WorkflowOperationException("No encoding profile was specified");
+
+    AttachmentSelector attachmentSelector = new AttachmentSelector();
+    for (String sourceFlavor : asList(sourceFlavorsOption)) {
+      attachmentSelector.addFlavor(sourceFlavor);
+    }
+    for (String sourceFlavor : asList(sourceFlavorOption)) {
+      attachmentSelector.addFlavor(sourceFlavor);
+    }
+    for (String sourceTag : asList(sourceTagsOption)) {
+      attachmentSelector.addTag(sourceTag);
+    }
+
+    // Look for elements matching the tag
+    Collection<Attachment> sourceElements = attachmentSelector.select(mediaPackage, tagsAndFlavors);
+
+    Map<Job, Attachment> jobs = new Hashtable<>();
+    try {
+      for (Attachment sourceElement : sourceElements) {
+        Job job = composerService.convertImage(sourceElement, profiles.toArray(new String[profiles.size()]));
+        jobs.put(job, sourceElement);
+      }
+      if (!waitForStatus(jobs.keySet().toArray(new Job[jobs.size()])).isSuccess()) {
+        throw new WorkflowOperationException("At least one image conversation job does not succeeded.");
+      }
+      for (Map.Entry<Job, Attachment> jobEntry : jobs.entrySet()) {
+        Job job = jobEntry.getKey();
+        Attachment sourceElement = jobEntry.getValue();
+        List<Attachment> targetElements =
+                (List<Attachment>) MediaPackageElementParser.getArrayFromXml(job.getPayload());
+        for (Attachment targetElement : targetElements) {
+          String targetFileName = PathSupport.toSafeName(FilenameUtils.getName(targetElement.getURI().getPath()));
+          URI newTargetElementUri = workspace.moveTo(targetElement.getURI(), mediaPackage.getIdentifier().compact(),
+                  targetElement.getIdentifier(), targetFileName);
+          targetElement.setURI(newTargetElementUri);
+          targetElement.setChecksum(null);
+
+          // set flavor on target element
+          if (targetFlavor != null) {
+            targetElement.setFlavor(targetFlavor);
+            if (StringUtils.equalsAny("*", targetFlavor.getType())) {
+              targetElement.setFlavor(MediaPackageElementFlavor.flavor(
+                      sourceElement.getFlavor().getType(), targetElement.getFlavor().getSubtype()));
+            }
+            if (StringUtils.equalsAny("*", targetFlavor.getSubtype())) {
+              targetElement.setFlavor(MediaPackageElementFlavor.flavor(
+                      targetElement.getFlavor().getType(), sourceElement.getFlavor().getSubtype()));
+            }
+          }
+          // set tags on target element
+          List<String> fixedTags = new ArrayList<>();
+          List<String> additionalTags = new ArrayList<>();
+          List<String> removingTags = new ArrayList<>();
+          for (String targetTag : asList(targetTagsOption)) {
+            if (!StringUtils.startsWithAny(targetTag, "+", "-")) {
+              if (additionalTags.size() > 0 || removingTags.size() > 0) {
```

Comment:
I would move this logic up into the WOH startup rather than waiting until after the jobs have completed to check this.  The behaviour is undefined at this point - if I were an admin reading this I would be unsure of what's going to happen.  IMO it's better to check this at the beginning, then throw an exception in this case.

Reply:
Agree, this logic should be unified on all workflow handlers IMO. But not with this PR.